src/v/kafka/server/server.cc

```diff
@@ -1292,43 +1302,174 @@ delete_topics_handler::handle(request_context ctx, ss::smp_service_group) {
     request.decode(ctx.reader(), ctx.header().version);
     log_request(ctx.header(), request);
 
-    auto unauthorized_it = std::partition(
-      request.data.topic_names.begin(),
-      request.data.topic_names.end(),
-      [&ctx](const model::topic& topic) {
-          return ctx.authorized(security::acl_operation::remove, topic);
-      });
+    // Determine the names and IDs that have been provided, and detect
+    // duplicates.
+    chunked_hash_set<model::topic> provided_names;
+    chunked_hash_set<model::topic> duplicate_provided_names;
+    chunked_hash_set<model::topic_id> provided_ids;
+    chunked_hash_set<model::topic_id> duplicate_provided_ids;
+
+    constexpr auto add_provided =
+      [](auto& provided, auto& duplicates, const auto& t) {
+          if (duplicates.contains(t) || !provided.insert(t).second) {
+              provided.erase(t);
+              duplicates.insert(t);
```

Comment:
TBH this function seems surprisingly complicated. i think it would benefit from either a comment explaining the semantics, or using more code to make it more self explanatory.